design/modular-development.md

```diff
@@ -0,0 +1,288 @@
+
+- Large index files to export everything.
+- Not as easily tree shakable to produce smaller bundle sizes.
+- Many methods on the client will never be used, thus taking up valuable space.
+
+## Modular Design
+
+We can create a more ala carte experience where the customer only pulls in the methods they need.  With the Azure SDK for JavaScript, we can create a client previously using the `ServiceClient` pattern:
+
+```typescript
+import { NotificationHubsClient } from "@azure/notification-hubs";
+
+const client = new NotificationHubsClient("<connection-string>", "<hub-name>");
+
+const registrationId = await client.createRegistrationId();
+```
+
+For a more modular design, we have a ServiceClient exposed as an interface with the minimal amount of things to communicate with the backend, and then this is passed into each method.
+
+```typescript
+import { createClientContext, createRegistrationId } from "@azure/notification-hubs/api";
+
+const context = createClientContext("<connection-string>", "<hub-name>");
+
+const registrationId = await createRegistrationId(context);
+```
+
+The pros of this design are as follows:
+
+- Fine grained exports to only give the customer what they need
+- Smaller bundle sizes the customer for exports.
+- Tree shaking is easier for bundlers.
+
+The cons of the modular design is the following:
+
+- Large index files to export everything.
+
+## Solution
+
+The goals of enabling both the class based `ServiceClient` and modular development are the following:
+
+- **Do not introduce breaking changes**
+- Keep the existing CommonJS compatibility.
+- Do not pollute the top level index with all exports, keeping the top level index.ts as is
+- Expose exports at "namespaces" or export subpaths
+
+## Subpath Exports
+
+To accomplish this goal, we will use [subpath exports](https://nodejs.org/api/packages.html#subpath-exports) which have been available since Node.js version 12.7.  This way, we can specify which files are exposed at which path to make a better experience for importing modules and not clutter the main `index.ts` file.
+
+Node added the following support for subpath exports in the following versions:
+
+- `v12.20.0` - Subpath `"exports"` patterns
+- `v12.16.0` - Unflag self-referencing a package using its name
+- `v12.7.0` - Introduce `"exports"` package.json field as a more powerful alternative to the classic `"main"` field
+- `v12.0.0` - Add support for ES modules using `.js` file extension via `package.json` `"type"` field.
+
+For subpaths, we can specify the path which is either a path or a wildcard path.  Inside the path, we can specify following:
+
+- `types` - The TypeScript types for that path
+- `require` - Common JS entry point
+- `import` - ES Modules entry point
+
+For example, we could have our client and associated methods exposed as `@azure/notification-hubs/api` and our models at `@azure/notification-hubs/models`.
+
+```json
+  "exports": {
+    ".": {
+      "types": "./types/src/index.js",
```

Comment:
I have same question with Jeff, do we export `src/api/index.js` & `/src/models/index.js` in `src/index.js`?

If yes does that mean we could have following two ways to import:

```
## Option 1
import { FooModel } from "@azure/notification-hubs/models";

## Option 2
import { FooModel } from "@azure/notification-hubs";
```

Reply:
No, there is no reason to pollute the main index with these exports.  This is a different experience where the current system relies on a really overloaded index.js.  This new experience is only going to be through ESM at the moment unless we somehow want to create a new CJS build with all the files like we do for ESM

Reply:
But yes, there will be two ways of importing the models, but if you're going to do the subpath, we would recommend Option 1